### PR TITLE
Rewrite MemoryAnalysis.md with up to date information.

### PR DESCRIPTION
**Pull request by JF002**

### CODE REVIEW

File: doc/MemoryAnalysis.md [hunk 1]
```diff
@@ -1,78 +1,284 @@
...
-## FreeRTOS heap and task stack
-FreeRTOS statically allocate its own heap buffer in a global variable named `ucHeap`. This is an array of *uint8_t*. Its size is specified by the definition `configTOTAL_HEAP_SIZE` in *FreeRTOSConfig.h*
-FreeRTOS uses this buffer to allocate memory for tasks stack and all the RTOS object created during runtime (timers, mutexes,...).
+The PineTime is equipped with the following memories:
+ - The internal RAM : **64KB**
+ - The internal Flash : **512KB**
+ - The external (SPI) Flash : **4MB**
 
-The function `xPortGetFreeHeapSize()` returns the amount of memory available in this *ucHeap* buffer. If this value reaches 0, FreeRTOS runs out of memory.
+Note that the NRF52832 cannot execute code stored in the external flash : we need to store the whole firmware in the internal flash memory, and use the external one to store graphicals assets, fonts,...
```
Comment (Itai-Nelken) [inline doc/MemoryAnalysis.md:7]:
```suggestion
Note that the NRF52832 cannot execute code stored in the external flash: we need to store the whole firmware in the internal flash memory, and use the external one to store graphicals assets, fonts...
```

File: doc/MemoryAnalysis.md [hunk 2]
```diff
@@ -1,78 +1,284 @@
...
+
+This document describes how the RAM and Flash memories and used in InfiniTime and how to analyze and monitor their usage. It was written in the context of [this memory analysis effort](https://github.com/JF002/InfiniTime/issues/313).
+
+## Code sections
+A binary is composed of multiple sections. Most of the time, these sections are .text, .rodata, .data and .bss but more sections can be defined in the linker script.
```
Comment (Itai-Nelken) [inline doc/MemoryAnalysis.md:12]:
```suggestion
A binary is composed of multiple sections. Most of the time, these sections are:
.text, .rodata, .data, and .bss
but more sections can be defined in the linker script.
```

File: doc/MemoryAnalysis.md [hunk 3]
```diff
@@ -1,78 +1,284 @@
...
+Here is a small description of these sections and where they end up in memory:
+
+ - **TEXT** = code (FLASH)
+ - **RODATA** = constants (FLASH)
+ - **DATA** = initialized variables (FLASH + RAM)
+ - **BSS** = uninitialized variables (RAM)
+
+
+## Internal FLASH
+The internal flash memory stores the whole firmware : code, variable that are not default-initialized, constants,...
```
Comment (Itai-Nelken) [inline doc/MemoryAnalysis.md:23]:
```suggestion
The internal flash memory stores the whole firmware: code, variable that are not default-initialized, constants...
```

File: doc/MemoryAnalysis.md [hunk 4]
```diff
@@ -1,78 +1,284 @@
...
+
+In this analysis, I used [Linkermapviz](https://github.com/PromyLOPh/linkermapviz).
+
+### Linkermapviz
+
+[Linkermapviz](https://github.com/PromyLOPh/linkermapviz) parses the MAP file and displays its content in a graphical way into an HTML page:
+
+![linkermapviz](./memoryAnalysis/linkermapviz.png)
+
+Using this tool, you can easily see the size of each symbol relative to the other one, check what is using most of the space,...
```
Comment (Itai-Nelken) [inline doc/MemoryAnalysis.md:39]:
```suggestion
Using this tool, you can easily see the size of each symbol relative to the other one, and check what is using most of the space...
```

File: doc/MemoryAnalysis.md [hunk 5]
```diff
@@ -1,78 +1,284 @@
...
+
+Also, as Linkermapviz is written in Python, you can easily modify it to adapt it to your firmware, export data in another format,... For example, [I modified it to parse the content of the MAP file and export it in a CSV file](https://github.com/JF002/InfiniTime/issues/313#issuecomment-842338620). I could later on open this file in LibreOffice Calc and use sort/filter functionality to search for specific symbols in specific files,...
```
Comment (Itai-Nelken) [inline doc/MemoryAnalysis.md:41]:
```suggestion
Also, as Linkermapviz is written in Python, you can easily modify it to adapt it to your firmware, export data in another format... For example, [I modified it to parse the contents of the MAP file and export it in a CSV file](https://github.com/JF002/InfiniTime/issues/313#issuecomment-842338620). I could later on open this file in LibreOffice Calc and use sort/filter functionality to search for specific symbols in specific files...
```

File: doc/MemoryAnalysis.md [hunk 6]
```diff
@@ -1,78 +1,284 @@
...
+
+### Puncover
+[Puncover](https://github.com/HBehrens/puncover) is another useful tools that analyses the binary file generated by the compiler (the .out file that contains all debug information). It provides valuable information about the symbols (data and code) : name, position, size, max stack of each functions, callers, callees,...
```
Comment (Itai-Nelken) [inline doc/MemoryAnalysis.md:44]:
```suggestion
[Puncover](https://github.com/HBehrens/puncover) is another useful tools that analyses the binary file generated by the compiler (the .out file that contains all debug information). It provides valuable information about the symbols (data and code): name, position, size, max stack of each functions, callers, callees...
```

File: doc/MemoryAnalysis.md [hunk 7]
```diff
@@ -1,78 +1,284 @@
...
+![Puncover](./memoryAnalysis/puncover.png)
+
+Puncover is really easy to install:
 
+ - clone the repo and cd into the cloned directory
+ - setup a venv
+    - `python -m virtualenv venv` 
+    - `source venv/bin/activate`
+ - Install : `pip install .`
+ - Run : `puncover --gcc_tools_base=/home/jf/nrf52/gcc-arm-none-eabi-9-2020-q2-update/bin/a-none-eabi- --elf_file /home/jf/git/PineTime/cmake-build-release/src/pinetime-app-1.1.0.out --src_root /home/jf/git/PineTime --build_dir /home/jf/git/PineTime/cmake-build-release/ `
```
Comment (Itai-Nelken) [inline doc/MemoryAnalysis.md:54]: The path's need to be made more "universal" in my opinion, but I'm not sure how.
Something like this:
*luse `/path/to/folder-or-file` where the path is, abd add a note to change it to the real path.

File: doc/MemoryAnalysis.md [hunk 8]
```diff
@@ -1,78 +1,284 @@
...
+This way, you can easily check what needs to be optimized : we should find a way to store big static data (like fonts and graphics) in the external flash memory, for example.
+
+It's always a good idea to check the flash memory space when working on the project : this way, you can easily check that your developments are using a reasonable amount of space. 
+
+### Links
+ - Analysis with linkermapviz : https://github.com/JF002/InfiniTime/issues/313#issuecomment-842338620
+ - Analysis with Puncover : https://github.com/JF002/InfiniTime/issues/313#issuecomment-847311392
+
+## RAM
+RAM memory contains all the data that can be modified at run-time: variables, stack, heap,...
```
Comment (Itai-Nelken) [inline doc/MemoryAnalysis.md:71]:
```suggestion
RAM memory contains all the data that can be modified at run-time: variables, stack, heap...
```

File: doc/MemoryAnalysis.md [hunk 9]
```diff
@@ -1,78 +1,284 @@
...
+
+### Data
+RAM memory can be *statically* allocated, meaning that the size and position of the data are known at compile-time:
+
+You can easily analyze the memory used by variables declared in the global scope using the MAP. You'll find them int the sections .BSS or .DATA. Linkermapviz and Puncover can be used to analyze their memory usage.
```
Comment (Itai-Nelken) [inline doc/MemoryAnalysis.md:76]:
```suggestion
You can easily analyze the memory used by variables declared in the global scope using the MAP. You'll find them in the .BSS or .DATA sections. Linkermapviz and Puncover can be used to analyze their memory usage.
```

File: doc/MemoryAnalysis.md [hunk 10]
```diff
@@ -1,78 +1,284 @@
...
+
+Variables declared in the scope of a function will be allocated on the stack. It means that the stack usage will vary according to the state of the program, and cannot be easily analyze at compile time.
```
Comment (Itai-Nelken) [inline doc/MemoryAnalysis.md:78]:
```suggestion
Variables declared in the scope of a function will be allocated on the stack. It means that the stack usage will vary according to the state of the program, and cannot be easily analyzed at compile time.
```

File: doc/MemoryAnalysis.md [hunk 11]
```diff
@@ -1,78 +1,284 @@
...
-In *src/CMakeLists.txt*:
+You can then display the file using objdump:
 
 ```
-set_target_properties(${EXECUTABLE_NAME} PROPERTIES
-        ...
-        LINK_FLAGS "-Wl,-wrap,malloc ..."
-        ...
-        )
+hexdump ram.bin -v  |less
```
Comment (Itai-Nelken) [inline doc/MemoryAnalysis.md:145]:
```suggestion
hexdump ram.bin -v  | less
```

File: doc/MemoryAnalysis.md [hunk 12]
```diff
@@ -1,78 +1,284 @@
...
 
-In *main.cpp*:
+#### Analysis 
+According to my experimentations, we don't use the stack that much, and 8192 bytes is probably way too big for InfiniTime!
+
+#### Links
+ - https://github.com/JF002/InfiniTime/issues/313#issuecomment-851035070
+
+### Heap
+The heap is declared in the [linker script](https://github.com/JF002/InfiniTime/blob/develop/nrf_common.ld#L136) and its current size is 8192 bytes. The heap is used for dynamic memory allocation(`malloc()`, `new`,...).
```
Comment (Itai-Nelken) [inline doc/MemoryAnalysis.md:198]:
```suggestion
The heap is declared in the [linker script](https://github.com/JF002/InfiniTime/blob/develop/nrf_common.ld#L136) and its current size is 8192 bytes. The heap is used for dynamic memory allocation(`malloc()`, `new`...).
```

File: doc/MemoryAnalysis.md [hunk 13]
```diff
@@ -1,78 +1,284 @@
...
+ - https://github.com/JF002/InfiniTime/issues/313#issuecomment-850890064
 
-#LittleVGL buffer
-*TODO*
 
-#NimBLE buffers
-*TODO*
+## FreeRTOS heap and task stack
+FreeRTOS statically allocate its own heap buffer in a global variable named `ucHeap`. This is an array of *uint8_t*. Its size is specified by the definition `configTOTAL_HEAP_SIZE` in *FreeRTOSConfig.h*
+FreeRTOS uses this buffer to allocate memory for tasks stack and all the RTOS object created during runtime (timers, mutexes,...).
```
Comment (Itai-Nelken) [inline doc/MemoryAnalysis.md:266]:
```suggestion
FreeRTOS uses this buffer to allocate memory for tasks stack and all the RTOS object created during runtime (timers, mutexes...).
```

**Review by MysteriousLog6 — CHANGES_REQUESTED** [review]
Replace `Flash memories and used in ` with `Flash memories are used in `

**Comment by JF002** [comment]
Thanks @Itai-Nelken and @MysteriousLog6 :+1:

**Pull request merged**